### PR TITLE
Fixed WSL cuda's OOM error

### PR DESCRIPTION
**Pull request by JoelSeniorLiang**
In WSL, the CUDA has limitation on the pinned memory. It's prohibited to allocate a large size of CUDA memory. But it can be bypassed in the program.

### CODE REVIEW

File: ggml-cuda.cu [hunk 1]
```diff
@@ -414,6 +415,42 @@ static cudaStream_t g_cudaStreams2[GGML_CUDA_MAX_STREAMS] = { nullptr };
...
+    fprintf(stderr, "Maximum amount of shared memory per SM:    %g KB\n",
+        prop.sharedMemPerMultiprocessor / 1024.0);
+    fprintf(stderr, "Maximum number of registers per block:     %d K\n",
+        prop.regsPerBlock / 1024);
+    fprintf(stderr, "Maximum number of registers per SM:        %d K\n",
+        prop.regsPerMultiprocessor / 1024);
+    fprintf(stderr, "Maximum number of threads per block:       %d\n",
+        prop.maxThreadsPerBlock);
+    fprintf(stderr, "Maximum number of threads per SM:          %d\n",
+        prop.maxThreadsPerMultiProcessor);
```
Comment (JohannesGaessler) [inline ggml-cuda.cu:453]: I think these info prints are excessive for default behavior. On a related note: I'm currently working on a PR for multi GPU support that also adds an info print upon initialization. It looks something like this:

```
ggml_init_cublas: found 2 CUDA devices:
  1. NVIDIA GeForce GTX 1070
  2. NVIDIA GeForce GTX 1050 Ti
```

Comment (liangmanlai) [reply ggml-cuda.cu:453]: Yes, those info prints are not that necessary.
Looking forward to the multi GPU support.

**Comment by JohannesGaessler** [comment]
I just opened a large PR for multi GPU support: https://github.com/ggerganov/llama.cpp/pull/1607

File: ggml-cuda.cu [hunk 2]
```diff
@@ -442,10 +479,17 @@ void * ggml_cuda_host_malloc(size_t size) {
     void * ptr = nullptr;
     cudaError_t err = cudaMallocHost((void **) &ptr, size);
     if (err != cudaSuccess) {
+        // The allocation error can be bypassed. A null ptr will assigned out of this function.
+        // This can fixed the OOM error in WSL.
+        cudaGetLastError();
```
Comment (howard0su) [inline ggml-cuda.cu:484]: why call this function here? You already have err?

File: ggml-cuda.cu (hunk 1 above)
Comment (howard0su) [reply ggml-cuda.cu:453]: Suggest to remove these

File: ggml-cuda.cu [hunk 3]
```diff
@@ -197,6 +197,7 @@ static __global__ void dequantize_block(const void * vx, float * y, const int k)
     // dequantize
     float & v0 = y[iybs + iqs + 0];
     float & v1 = y[iybs + iqs + y_offset];
+
```
Comment (howard0su) [inline ggml-cuda.cu:200]: cleanup

File: ggml-cuda.cu [hunk 4]
```diff
@@ -442,10 +479,17 @@ void * ggml_cuda_host_malloc(size_t size) {
...
     cudaError_t err = cudaMallocHost((void **) &ptr, size);
     if (err != cudaSuccess) {
+        // The allocation error can be bypassed. A null ptr will assigned out of this function.
+        // This can fixed the OOM error in WSL.
+        cudaGetLastError();
         fprintf(stderr, "WARNING: failed to allocate %.2f MB of pinned memory: %s\n",
             size/1024.0/1024.0, cudaGetErrorString(err));
         return nullptr;
     }
+    else{
```
Comment (howard0su) [inline ggml-cuda.cu:489]: add a space between else and {

File: ggml-cuda.cu [hunk 5]
```diff
@@ -442,10 +479,17 @@ void * ggml_cuda_host_malloc(size_t size) {
...
+        fprintf(stderr, "INFO: succeed to allocate %.2f MB of pinned memory\n",
```
Comment (howard0su) [inline ggml-cuda.cu:490]: do we really need this? it is not a meaningful info for most users.

File: ggml-cuda.cu (hunk 2 above)
Comment (liangmanlai) [reply ggml-cuda.cu:484]: Failed to allocate memory error.

File: ggml-cuda.cu (hunk 5 above)
Comment (liangmanlai) [reply ggml-cuda.cu]: codes are cleaned up.

File: ggml-cuda.cu (hunk 2 above)
Comment (howard0su) [reply ggml-cuda.cu:447]: sorry, i mean why we need to call cudaGetLastError and also discard the return code?

Comment (liangmanlai) [reply ggml-cuda.cu:447]: Yes, just ignore the cuda error. In llama_ctx_buffer.resize(), it will fall back to pageable memory if it's encountering errors.

**Review by slaren — APPROVED** [review]

**Comment by slaren** [comment]
This is good, this was the intended behavior when the host malloc fails, and not clearing the error was an oversight.

**Pull request merged**